### DISCUSSION TITLE
Replacing pipeline components with different models

### DISCUSSION THREAD

**Discussion by yumikim381** [deepdoctection/deepdoctection#302] [General]
is it possible to replace pipeline components with models that aren't specified in ModelCatalog?  
For example to do text extraction service to PaddleOCR instead of Tesseract?
More generally can you also register any kinds of models to model catalog by using register() method from deepdoctection/extern
/model.py?

**Comment by yumikim381** [comment]
Also one other question: could you tell me how small the ROIs (input to text extraction service ) generally is using dd.analyzer's layout model? Are they as small as work or paragraph or list etc?

**Reply by JaMe76** [reply]
I am not quite sure about your question. But there is no requirement about the area of any detection result.

**Reply by yumikim381** [reply]
So the input to OCR can be as small as a single character and as big as an entire pdf page?

**Reply by JaMe76** [reply]
Theoretically yes. There is no restriction with respect to the OCR result. However, to get a full working end-to-end engine you will likely change many other default parameters and even change the processing structure. 

Take grouping characters to words: There is no such logic in there yet. 
You will also likely have to change the text ordering process if you assume much larger/smaller OCR outputs.

**Comment by JaMe76** [comment]
> is it possible to replace pipeline components with models that aren't specified in ModelCatalog? For example to do text extraction service to PaddleOCR instead of Tesseract? More generally can you also register any kinds of models to model catalog by using register() method from deepdoctection/extern /model.py?

It does not work out-of-the box. If you want to use a particular library you have to write a `deepdoctection` interface, for otherwise the library will not know how to invoke the model and how to parse the returned results. 

That is, if you want to use an end-to-end OCR predictor from PaddleOCR, you will have to wrap the ocr detector in a deepdoctection `ObjectDetector` interface:

```python

class PaddleOCRDetector(ObjectDetector)

def __init__(self, config_path_yaml, path_weights):  # if the paddle model requires a config file and a weights file
    self.name = "paddle-ocr"
    self.config = config_path_yaml
    self. path_weights = path_weights
    self.paddle_model = # code to instantiate the PaddleOCR model

def predict(self. np_img: ImageType):
    # transform the numpy image so that it can be loaded into the paddle model
    paddle_input = transform_to_paddle_input(np_image)
    paddle_outputs = self.paddle_model(paddle_input)
   # transform paddle outputs into a list of 'DetectionResult` 
   detection_results = paddle_outputs_to_detection_results(paddle_outputs)
   return detection_results

def get_requirements(cls):
  return [] # or you can write a requirement function, to check if PaddlePaddle is installed

def clone()
    return self.__class__(# yout init input values)
```

I recommend to look into some examples in the library, how the inferface a implemented, e.g. (`deepdoctection.extern.doctr`, `deepdoctection.extern.tessocr`) It obviously depends onthe inner model to fill in the details.

You can then plug your wrapper in the `TextExtractionService`.

**Reply by yumikim381** [reply]
Thank you so much !! This is very helpful. Thanks again for publishing your great work
